OptimizelySDK/Odp/OdpEventManager.cs

```diff
@@ -0,0 +1,451 @@
+            {
+                _logger.Log(LogLevel.ERROR, Constants.ODP_INVALID_DATA_MESSAGE);
+                return;
+            }
+
+            odpEvent.Data = AugmentCommonData(odpEvent.Data);
+            if (!_eventQueue.TryAdd(odpEvent))
+            {
+                _logger.Log(LogLevel.WARN, "Payload not accepted by the queue.");
+            }
+        }
+
+        /// <summary>
+        /// Adds an event to the current batch being created 
+        /// </summary>
+        /// <param name="odpEvent"></param>
+        private void AddToBatch(OdpEvent odpEvent)
+        {
+            lock (_mutex)
+            {
+                // Reset the deadline if starting a new batch.
+                if (_currentBatch.Count == 0)
+                {
+                    _flushingIntervalDeadline = DateTime.Now.MillisecondsSince1970() +
+                                                (long)_flushInterval.TotalMilliseconds;
+                }
+
+                _currentBatch.Add(odpEvent);
+
+                if (_currentBatch.Count >= _batchSize)
+                {
+                    FlushQueue();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Ensures queue processing is stopped marking this instance as disposed 
+        /// </summary>
+        public void Dispose()
+        {
+            if (Disposed)
+            {
+                return;
+            }
+
+            Stop();
+            Disposed = true;
+        }
+
+        /// <summary>
+        /// Associate a full-stack userid with an established VUID
+        /// </summary>
+        /// <param name="userId">Full-stack User ID</param>
+        public void IdentifyUser(string userId)
+        {
+            var identifiers = new Dictionary<string, string>
+            {
+                {
+                    OdpUserKeyType.FS_USER_ID.ToString(), userId
+                },
+            };
+
+            var odpEvent = new OdpEvent(Constants.ODP_EVENT_TYPE, "identified", identifiers);
+            SendEvent(odpEvent);
+        }
+
+        /// <summary>
+        /// Update ODP configuration settings
+        /// </summary>
+        /// <param name="odpConfig">Configuration object containing new values</param>
+        public void UpdateSettings(OdpConfig odpConfig)
+        {
+            _odpConfig.Update(odpConfig.ApiKey, odpConfig.ApiHost, odpConfig.SegmentsToCheck);
+        }
+
+        /// <summary>
+        /// ODP event data has invalid types
+        /// </summary>
+        /// <param name="data">Data to be analyzed</param>
+        /// <returns>True if a type is not a valid type or is not null otherwise False</returns>
+        private bool InvalidDataFound(Dictionary<string, dynamic> data)
+        {
+            return data.Any(item =>
+                item.Value != null &&
+                !_validOdpDataTypes.Any(t =>
+                    t.Equals(item.Value.GetType().Name, StringComparison.OrdinalIgnoreCase))
```

Comment:
We can consider to improve this. For a long data set and valid type set, this can be slow for every event.